# Problem Set 02: Python Basics I

**Follows Lecture 02**

Full solutions are provided in the separate `A1.py`-`A4.py` files, but first try to solve each problem using only the hints.

This problem set covers atomic types, type conversion, arithmetic and logical operators, lists, copying, dictionaries, and conditionals. Each exercise can be solved using only material from Lecture 02.

Use descriptive variable names and label the values you print. The collapsed answer sections give hints and numerical checks, but not complete solution code. Write and run your own solution before opening them.

## 1. Prices and Purchasing Power

A worker's weekly wage rises between two years. Over the same period, the price index also rises.

1. Print `weekly_wage_last_year_text` and its type.
2. Convert it to a float called `weekly_wage_last_year`.
3. Calculate the percentage growth in the weekly wage.
4. Calculate the inflation rate as the percentage growth in the price index.
5. Calculate the real weekly wage in each year by dividing the weekly wage by the price index and multiplying by 100.
6. Calculate the percentage growth in the real weekly wage.
7. Store the comparison `real_weekly_wage_this_year > real_weekly_wage_last_year` as `purchasing_power_increased`.
8. Print the results with clear labels.

In one sentence, explain why purchasing power changes by less than the nominal wage.

In [16]:
weekly_wage_last_year_text = "4800.0"
weekly_wage_this_year = 4992.0

price_index_last_year = 100.0
price_index_this_year = 103.0

In [ ]:
# print solution from python script
# normally we would just use:
# import A1
# the import will run all of the code in A1.py. But since variables like weekly_wage_last_year_text is not defined in A1.py, the import would fail. 
# instead we can use the %run command to run the script in the current namespace, so that it has access to all variables defined in this notebook.
# you can use this to access the solutions directly from the notebook in the future
%run -i "python_scripts/A1.py"
# if you keep the solutions in the same folder as your notebook, you can just use %run -i "A1.py" instead of the full path.


# the solution is self explanatory, otherwise come ask me:)

Original wage: 4800.0
Original type: <class 'str'>
Nominal wage growth: 4.0000000000000036 percent
Inflation rate: 3.0000000000000027 percent
Real wage last year: 4800.0
Real wage this year: 4846.601941747573
Real wage growth: 0.9708737864077666 percent
Purchasing power increased: True


**Explanation:**

<details>
<summary><strong>Hints and Results</strong></summary>

- Use `type(...)` to inspect the original wage and `float(...)` to convert it.
- Percentage growth is the change divided by the original value, multiplied by 100.
- The nominal wage growth rate is `4.0` percent.
- The inflation rate is `3.0` percent.
- The real weekly wages are `4800.0` and approximately `4846.60`.
- Real wage growth is approximately `0.97` percent.
- `purchasing_power_increased` should be `True`.

The nominal wage rises by 4 percent, but prices rise by 3 percent, so most of the wage increase compensates for higher prices.

</details>

## 2. A Household Budget After an Energy Price Increase

The lists below describe a household's monthly spending. The government wants to study a counterfactual in which energy spending rises by 25 percent while all other spending stays unchanged.

1. Print the number of spending categories.
2. Create `higher_energy_spending` as an independent copy of `baseline_spending`.
3. Increase the energy entry in the copied list by 25 percent. Do not type the new value directly.
4. Create a dictionary called `budget_scenario` with the keys `"categories"`, `"baseline"`, and `"higher_energy"` and the corresponding lists as values.
5. Use the dictionary to print the name of the affected category and its spending in both scenarios.
6. Calculate the total spending in each scenario and the extra monthly spending caused by the energy-price increase.
7. Print `baseline_spending` again to confirm that it has not changed.

In one sentence, explain why copying the list matters for this comparison.

In [59]:
spending_categories = ["Rent", "Food", "Energy", "Transport"]
baseline_spending = [9000.0, 3600.0, 1200.0, 950.0]

In [ ]:
# Solution with explanation

# 1.
print(f'There are {len(spending_categories)} spending categories') #len() returns the number of elements in a list

# 2.
higher_energy_spending = baseline_spending[:] 
# create a copy of the baseline_spending list. [:] is a slice that returns all elements of the list, so we can use it to create a copy of the list, 
# by setting the new variable to be equal to each element in the old list, but not the old list itself.
# if we just did higher_energy_spending = baseline_spending, we would just create a reference to the same list, 
# so if we change one, the other would change as well.
# alternatively use baseline_spending.copy() to create a copy of the list.

# 3.
higher_energy_spending[2] *= 1.25  # multiply the energy element in the list by 1.25 to increase energy spending by 25%.

# 4.
budget_scenario = dict(categories=spending_categories, baseline=baseline_spending, higher_energy=higher_energy_spending) 
# create a dictionary with the spending categories and the two spending scenarios. Before the '=' are the keys and after are the values.

# 5.
print(f'The baseline spending on {budget_scenario["categories"][2]} is {budget_scenario["baseline"][2]:_.0f} \
and the higher energy spending is {budget_scenario["higher_energy"][2]:_.0f}')
# '\' is a continuation character that allows us to split a long line of code into multiple lines for better readability.
# we can access the values in the dictionary using the keys, and we can access the elements in the lists using their index.
# we use f-strings to format the output, which allows us to include variables in the string by enclosing them in curly braces {}.
# we format the numbers using ':_.0f', to use a thousands separator and no decimal places. 

# 6.
print(f'The total baseline spending is {sum(budget_scenario["baseline"]):_.0f} and the total higher energy spending is \
{sum(budget_scenario["higher_energy"]):_.0f}, so the total spending increased by {sum(budget_scenario["higher_energy"]) - sum(budget_scenario["baseline"]):_.0f}')

# 7.
print(baseline_spending) # it did not change

There are 4 spending categories
The baseline spending on Energy is 1_200 and the higher energy spending is 1_500
The total baseline spending is 14_750 and the total higher energy spending is 15_050, so the total spending increased by 300
[9000.0, 3600.0, 1200.0, 950.0]


<details>
<summary><strong>Hints and Results</strong></summary>

- Use `.copy()` before changing the energy entry.
- Energy is at index `2` in both lists.
- Use `sum(...)` for each total and square brackets to access values in the dictionary.
- There are `4` spending categories.
- Energy spending rises from `1200.0` DKK to `1500.0` DKK.
- Baseline total spending is `14750.0` DKK.
- Total spending with higher energy costs is `15050.0` DKK.
- Extra monthly spending is `300.0` DKK.
- The original list must still be `[9000.0, 3600.0, 1200.0, 950.0]`.

Copying creates a separate list, so changing the counterfactual does not erase the baseline used for comparison.

</details>

## 3. Two Quarters of Falling GDP

The list below contains quarterly real GDP growth rates in percent. For this programming exercise, call the economy a *technical recession* when GDP growth is negative in the final two quarters. This simple rule is not a complete economic assessment of a recession.

1. Use negative indexes to print the labels and growth rates for the final two quarters.
2. Store a Boolean called `technical_recession` that is `True` only if both final growth rates are negative. Use `and`.
3. Write one `if` / `elif` / `else` block that prints:
   - `Two consecutive quarters of falling GDP` if both final growth rates are negative,
   - `GDP fell in the latest quarter` if only the latest growth rate is negative,
   - `GDP did not fall in the latest quarter` otherwise.
4. Run the code with the original data.
5. Change Q3 growth to `0.1` and rerun it.
6. Then change Q4 growth to `0.2` and rerun it again.

Record the message printed in each case.

In [79]:
# store lists in same cell so the cell always returns the same output, even if the list is changed as the code is run cell by cell
quarter_labels = ["Q1", "Q2", "Q3", "Q4"]
quarterly_gdp_growth = [0.6, 0.2, -0.3, -0.4]

# 1.
print('q1-q4')
print(f'The real GDP growth was {quarterly_gdp_growth[-2]} in {quarter_labels[-2]} and {quarterly_gdp_growth[-1]} in {quarter_labels[-1]}')

# 2.
technical_recession = quarterly_gdp_growth[-2] < 0 and quarterly_gdp_growth[-1] < 0 # we use and to check if both conditions are true

# 3.

if technical_recession: #since technical_recession is a boolean variable, we can use it directly in the if statement
    print('Two consecutive quarters of falling GDP')
elif quarterly_gdp_growth[-1] < 0: #elif - "else if" - checks the second condition if the first is false
    print('GDP fell in the latest quarter')
else: # if both conditions are false, we can use else to catch all other cases
    print('GDP did not fall in the latest quarter')

# 4. this is 1-3 in the output
# create a function that does the same as 1-3
def check_gdp_growth(gdp_growth):
    print(f'The real GDP growth was {gdp_growth[-2]} in {quarter_labels[-2]} and {gdp_growth[-1]} in {quarter_labels[-1]}')
    technical_recession = gdp_growth[-2] < 0 and gdp_growth[-1] < 0
    if technical_recession:
        return print('Two consecutive quarters of falling GDP')
    elif gdp_growth[-1] < 0:
        return print('GDP fell in the latest quarter')
    else:
        return print('GDP did not fall in the latest quarter')

# 5.
print('q5')
quarterly_gdp_growth[-2] = 0.1
check_gdp_growth(quarterly_gdp_growth) # we use the function with quarterly_gdp_growth as the argument(input)

# 6.
print('q6')
quarterly_gdp_growth[-1] = 0.2
check_gdp_growth(quarterly_gdp_growth)

q1-q4
The real GDP growth was -0.3 in Q3 and -0.4 in Q4
Two consecutive quarters of falling GDP
q5
The real GDP growth was 0.1 in Q3 and -0.4 in Q4
GDP fell in the latest quarter
q6
The real GDP growth was 0.1 in Q3 and 0.2 in Q4
GDP did not fall in the latest quarter


**Messages:**

1. Original data:
2. After changing Q3 growth to `0.1`:
3. After changing Q4 growth to `0.2`:

<details>
<summary><strong>Hints and Results</strong></summary>

- The final two positions are `-2` and `-1`.
- Combine the two comparisons with `and`.
- With the original data, the final rates are `-0.3` and `-0.4`, `technical_recession` is `True`, and the first message is printed.
- After changing Q3 growth to `0.1`, `technical_recession` is `False`, and the second message is printed.
- After also changing Q4 growth to `0.2`, the third message is printed.

</details>

## 4. Find and Fix the Errors in an Inflation Calculation

The code below is intended to compare the cost of the same consumption basket across two years and classify the resulting inflation rate. It runs without an error message, but it contains three mistakes.

1. Find and fix all three mistakes.
2. Rerun the cell after each change.
3. Check that last year's basket is preserved, that the inflation formula uses the whole change in basket cost, and that the more specific inflation category can be reached.
4. Write one sentence explaining each correction.

In [83]:
# corrected code

basket_items = ["Food", "Energy", "Transport"]
last_year_costs = [500.0, 300.0, 200.0]

this_year_costs = last_year_costs[:]  # create a copy of the last_year_costs list to avoid modifying the original list
this_year_costs[1] = 330.0

last_year_total = sum(last_year_costs)
this_year_total = sum(this_year_costs)
inflation_rate = (this_year_total - last_year_total) / last_year_total * 100

if inflation_rate > 2:
    inflation_label = "Inflation above 2 percent"
elif inflation_rate > 0:
    inflation_label = "Positive inflation"
else:
    inflation_label = "No inflation"

print("Last year's costs:", last_year_costs)
print("This year's costs:", this_year_costs)
print("Last year's total:", last_year_total)
print("This year's total:", this_year_total)
print("Inflation rate:", inflation_rate, "percent")
print("Classification:", inflation_label)

Last year's costs: [500.0, 300.0, 200.0]
This year's costs: [500.0, 330.0, 200.0]
Last year's total: 1000.0
This year's total: 1030.0
Inflation rate: 3.0 percent
Classification: Inflation above 2 percent


**Errors corrected:**

1. ***Reference to same place in memory***: `this_year_costs = last_year_costs` makes `this_year_costs` point to the same list as `last_year_costs`, so changing one changes the other. Use `.copy()` to create an independent copy. Or slice `[:]` to create a new list with the same elements.
2. ***Order of operations***: The inflation formula `this_year_total - last_year_total / last_year_total * 100` does not calculate the percentage change correctly because division and multiplication have higher precedence than subtraction. Use parentheses to ensure the entire change is divided by last year's total.
3. ***Order of conditional checks***: The `if` / `elif` / `else` block checks for positive inflation first, so the more specific category of 'inflation above 2 percent' is never reached. Check for the more specific category first.

<details>
<summary><strong>Hints and Results</strong></summary>

- One problem concerns two variable names referring to the same list. Make an independent copy.
- One problem concerns the brackets in the percentage-change formula. The full change belongs in the numerator.
- One problem concerns the order of the conditional branches. Test the higher threshold before the broader positive-inflation condition.
- The preserved last-year costs are `[500.0, 300.0, 200.0]`.
- This year's costs are `[500.0, 330.0, 200.0]`.
- The two totals are `1000.0` and `1030.0`.
- The correct inflation rate is `3.0` percent.
- The correct classification is `Inflation above 2 percent`.

</details>

Restart the kernel and run the notebook from top to bottom. Check that your code produces the listed numerical results and that changing an input changes every result that depends on it.